pkg/scheduler/actions/allocate/allocate.go

```diff
@@ -150,11 +150,32 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
 				glog.V(3).Infof("Considering Task <%v/%v> on node <%v>. Task request: <%v>; Idle: <%v>; Used: <%v>; Releasing: <%v>; Backfilled: <%v>",
 					task.Namespace, task.Name, node.Name, task.Resreq, node.Idle, node.Used, node.Releasing, node.Backfilled)
 
-				if task.InitResreq.LessEqual(node.GetAccessibleResource()) {
-					glog.V(3).Infof("Binding Task <%v/%v> to node <%v>",
-						task.Namespace, task.Name, node.Name)
+				// allow over-allocate for starving job 
+				isJobStarving := job.Starving(ssn.StarvationThreshold)
+				toOverAllocateForStarv := false
+				if isJobStarving {
+					netResource := node.Allocatable.Clone()
+					for _, nodeTask := range node.Tasks {
+						if nodeTask.Job == job.UID &&
+						(nodeTask.Status == api.OverOccupied || nodeTask.Job == job.UID && nodeTask.Status == api.Allocated) {
+							netResource.Sub(nodeTask.InitResreq)
+							glog.Infof("removing %v of task %s from node %s", nodeTask.InitResreq, nodeTask.Name, node.Name)
+						}
+					}
+
+					// toOverAllocateForStarv == true means when allocating for starving job, over-allocate resources ignoring 
+					// resources being used on the node to prevent other non-startving job from taking the resources
+					toOverAllocateForStarv = (isJobStarving &&
+					!task.InitResreq.LessEqual(node.GetAccessibleResource()) &&
+					task.InitResreq.LessEqual(netResource))
+				}
+
+				if task.InitResreq.LessEqual(node.GetAccessibleResource()) || toOverAllocateForStarv {
+					glog.V(3).Infof("Binding Task <%v/%v> to node <%v>", task.Namespace, task.Name, node.Name)
+
+					usingBackfillTaskRes := !task.InitResreq.LessEqual(node.Idle) && !isJobStarving
 
-					if err := ssn.Allocate(task, node.Name, !task.InitResreq.LessEqual(node.Idle)); err != nil {
+					if err := ssn.Allocate(task, node.Name, usingBackfillTaskRes, toOverAllocateForStarv); err != nil {
```

Comment:
Try not passing bool into ssn.Allocate. It starts to get ugly. :)

Reply:
being refactored